### PR TITLE
`imul`, `isqr` fixes

### PR DESCRIPTION
**Pull request by alexeykudinkin**
First, `isqr` isn't in-place in current implementation.
Making it in-place, however reveals that `imul` implementation (which in itself looks like a clone of `bigMulTo`) is flawed: therefore its replacement with single entry-point seems reasonable enough for me.

Cons?

### CODE REVIEW

**Comment by alexeykudinkin** [comment]
Holy grail, wrong base! Sorry for that.

**Comment by alexeykudinkin** [comment]
That way looks much better.

File: lib/bn.js [hunk 1]
```diff
@@ -1104,7 +1060,7 @@ BN.prototype.sqr = function sqr() {
 
 // `this` * `this` in-place
 BN.prototype.isqr = function isqr() {
-  return this.mul(this);
+  return this.imul(this.clone());
```
Comment (indutny) [inline lib/bn.js]: I agree with this, but disagree with above.

**Comment by indutny** [comment]
@alexeykudinkin is there any kind of regression test or anything that demonstrates `imul` problems? I'd rather fix current `imul` implementation.

**Comment by alexeykudinkin** [comment]
For what purpose there're separate implementations for `imul` & `bigMulTo`?

Regression: Patch `isqr` the way described & run `npm test`.

**Comment by indutny** [comment]
It is for performance reasons, it should not allocate anything.

`smallMulTo` is optimized for numbers with low amount of limbs, and is not correct for bigger numbers.

**Comment by indutny** [comment]
Argh, I know what's happening. It needs a `bigMulTo`. Let's ditch it. LGTM

**Comment by indutny** [comment]
Landed in b627b19, thank you!

**Pull request closed without merging**